api/controllers/oauth-controller.go

```diff
@@ -0,0 +1,154 @@
+package controllers
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/linkedin"
+
+	"github.com/gin-gonic/gin"
+)
+
+type OAuthController struct {
+}
+
+var (
+	state = "linkedin_random_csrf_string" // TODO needs more secure implementation
+)
+
+func (c *OAuthController) GetLinkedInAuthorizationUrl(ctx *gin.Context) {
+	ctx.JSON(200, gin.H{
+		"url": getConfig().AuthCodeURL(state),
+	})
+}
+
+func (c *OAuthController) OAuthCallback(ctx *gin.Context) {
+	query := ctx.Request.URL.Query()
+	codes := query["code"]
+	states := query["state"]
+	token, err := getConfig().Exchange(oauth2.NoContext, codes[0])
+	if err != nil {
+		ctx.JSON(400, gin.H{
+			"error": err.Error(),
+		})
+		return
+	}
+	client := getConfig().Client(oauth2.NoContext, token)
+	profile, err := getProfile(client, token.AccessToken)
+	if err != nil {
+		ctx.JSON(400, gin.H{
+			"error": err.Error(),
+		})
+		return
+	}
+	emailResponse, err := getEmail(client, token.AccessToken)
+	if err != nil {
+		ctx.JSON(400, gin.H{
+			"error": err.Error(),
+		})
+		return
+	}
+
+	ctx.JSON(200, gin.H{
+		"code":        codes[0],
+		"state":       states[0],
+		"token":       token,
+		"profile":     profile,
+		"first_name":  profile.FirstName.Localized["en_US"],
+		"last_name":   profile.LastName.Localized["en_US"],
+		"linkedin_id": profile.ID,
+		"email":       emailResponse.Elements[0].HandleValue.EmailAddress,
+	})
+}
+
+func getConfig() *oauth2.Config {
+	return &oauth2.Config{
+		ClientID:     os.Getenv("LINKEDIN_CLIENT_ID"),
+		ClientSecret: os.Getenv("LINKEDIN_CLIENT_SECRET"),
+		RedirectURL:  fmt.Sprintf("%s/oauth/linkedin_callback", os.Getenv("LINKEDIN_CALLBACK_HOST")),
+		Scopes:       []string{"r_emailaddress", "r_liteprofile"},
+		Endpoint:     linkedin.Endpoint,
+	}
+}
+
+func getProfile(client *http.Client, accessToken string) (LiteProfile, error) {
+	req, err := http.NewRequest("GET", "https://api.linkedin.com/v2/me?projection=(id,firstName,lastName,email,profilePicture(displayImage~:playableStreams))", nil)
```

Comment:
same here ... we can get the BASE_URL and the scopes or URL path separately.

Reply:
ditto